practices/appeng/adrs/0001-use-typescript.md

```diff
@@ -0,0 +1,99 @@
+# 0001 - Preference to use TypeScript in place of JavaScript
+
+**Status:** In progress
+
+**Date Accepted:** TBD
+
+**Reviewers:** @suzubara @jim @rswerve <tag yourself!>
+
+## Context
+
+When in the business of building software for the web, JavaScript is largely unavoidable. It is virtually the only language option for interacting with browser APIs in order to enhance user experience & interfaces, and implement modern features that meet the high expectations of users today. It is also rapidly gaining popularity as a backend language, with the advent of Node.JS, web frameworks such as Express and Next.js, and the growing need to better optimize UI code by rendering it on the server first.
+
+However, JavaScript has more than its share of quirks, and because it runs natively in browsers, it can be too easy to introduce bugs and errors that aren't realized until encountered by an end user. These days it is crucial to accompany JavaScript code with development tools such as linters and automated tests in order to avoid those scenarios. TypeScript is another such tool, [introduced by Microsoft](https://www.typescriptlang.org/why-create-typescript) as a superset of JavaScript that adds [type safety](https://en.wikipedia.org/wiki/Type_safety) to the language. By writing code in TypeScript instead of JavaScript, engineers can avoid entire categories of errors that would be caught well ahead of code being deployed to production.
+
+### Why type safety?
+
+The decision to use TypeScript assumes a consensus that type safety in a programming language is better than no type safety. Below are some reasons for this:
+
+- A strong static typing system can help reduce the possibility of bugs at runtime, since type errors are captured earlier, during compilation.
```

Comment:
- Types free the programmer from writing an entire class of tests. You can be sure that the objects being passed into your function have fields with the correct types, many runtime checks become unneccecary and many tests cases no longer need to be written
- The most useful type check is defining wether a variable or field is required or possibly undefined. The [billion-dollar mistake](https://en.wikipedia.org/wiki/Tony_Hoare#Apologies_and_retractions) was due to an insufficienlty expressive type system

A couple suggestions. I’m sure we can drum up more. ^